drivers/kubevirt/kubevirt.go

```diff
@@ -507,3 +515,313 @@ func AddPVCsToVirtualMachine(vm kubevirtv1.VirtualMachine, pvcs []*corev1.Persis
+		if pod.Labels["kubevirt.io"] == "virt-launcher" && pod.Status.Phase == corev1.PodRunning {
+			if found.Name != "" {
+				// there should be only one VM pod in the running state (otherwise live migration is in progress)
+				return nil, fmt.Errorf("more than 1 KubeVirt pods (%s, %s) are in running state for volume %s",
+					found.Name, pod.Name, vol.ID)
+			}
+			found = pod
+		}
+	}
+	if found.Name == "" {
+		return nil, fmt.Errorf("failed to find a running pod for volume %s", vol.ID)
+	}
+	return &found, nil
+}
+
+func IsVMBindMounted(virtualMachineCtx *scheduler.Context, driver scheduler.Driver, volumeDriver volume.Driver, initialCheck bool) (bool, error) {
+	vols, err := driver.GetVolumes(virtualMachineCtx)
+	if err != nil {
+		return false, err
+	}
+	if len(vols) == 0 {
+		return false, fmt.Errorf("no volumes found for VM [%s] in namespace [%s]", virtualMachineCtx.App.Key, virtualMachineCtx.App.NameSpace)
+	}
+
+	// Get the node where pod is scheduled
+	vms, err := GetAllVMsFromScheduledContexts([]*scheduler.Context{virtualMachineCtx})
+	if err != nil {
+		return false, err
+	}
+	if len(vms) != 1 {
+		return false, fmt.Errorf("expected 1 VM for VM [%s] in namespace [%s] but got [%d]", virtualMachineCtx.App.Key, virtualMachineCtx.App.NameSpace, len(vms))
+	}
+	vm := vms[0]
+	vmNodeName, err := GetNodeOfVM(vm)
+	if err != nil {
+		return false, err
+	}
+	log.Infof("VM [%s] is deployed on node [%s]", virtualMachineCtx.App.Key, vmNodeName)
+
+	// Keep a track of replicaset and consider this as the source of truth
+	volInspect, err := volumeDriver.InspectVolume(vols[0].ID)
+	if err != nil {
+		return false, fmt.Errorf("failed to inspect volume [%s]: %w", vols[0].ID, err)
+	}
+	globalReplicSet := volInspect.ReplicaSets
+	log.InfoD("Length of replicaset: %d", len(globalReplicSet))
+
+	// Check if the volume is mounted on the node
+	for _, vol := range vols {
+		// Check if the volume is mounted on the node
+		volInspect, err := volumeDriver.InspectVolume(vol.ID)
+		if err != nil {
+			return false, fmt.Errorf("failed to inspect volume [%s]: %w", vol.ID, err)
+		}
+		nodeIpAttachedOn := volInspect.AttachedOn
+		nodeNameAttachedOn, err := node.GetNodeByIP(nodeIpAttachedOn)
+		if err != nil {
+			return false, fmt.Errorf("failed to get node name by IP [%s]: %w", nodeIpAttachedOn, err)
+		}
+		log.Infof("Volume [%s] is attached on node [%s]", vol.ID, nodeNameAttachedOn.Name)
+		if nodeNameAttachedOn.Name != vmNodeName {
+			return false, fmt.Errorf("volume [%s] is not attached on node [%s]", vol.ID, vmNodeName)
+		}
+
+		// Check if all the replicas are attached to the same node
+		replicaset := volInspect.ReplicaSets
+
+		// check if the replicaset size is same as the global replicaset size
+		if len(replicaset) != len(globalReplicSet) {
+			return false, fmt.Errorf("replicaset count mismatch for volume [%s]", vol.ID)
+		}
+		// The criteria to call the bind mount successful is to check if the replicaset of all volumes should be same
```

Comment:
I still think that the collocation check should be in a separate function. That will simplify the logic here. For bind mount, it is sufficient to verify the output of "mount". No other check is needed.

Reply:
Moved volume colocation logic to another function `VerifyVolumeReplicaCollocation`.